### PR TITLE
fix(async-flow)!: stopgap `E` only for `makeReplayMembraneForTesting`

### LINKED ISSUE #9772: Disable stopgap E support in async-flow
## What is the Problem Being Solved?

#9519 added stop-gap support for `E` in anticipation of #9299. We realized that `E` will not be required for MVP orchestration use cases, so we should disable or remove the stop-gap support before releasing async-flow, and ensure nothing breaks.

## Description of the Design

Revert #9519 or equivalent disablement

## Security Considerations

None

## Scaling Considerations

None

## Test Plan

disable the async-flow eventual-send tests

verify that no existing orchestration test fails

## Upgrade Considerations

None

### PR DESCRIPTION
**Pull request by michaelfig**
closes: #9772

## Description
To accomplish #9772, disable the stopgap eventual send support except when explicitly calling `makeReplayMembraneForTesting({..., __eventualSendForTesting: true })`.  We do this surgically to allow the membrane tests to prevent regression concerning the stopgap eventual-send guest support, and provide coverage for when we implement proper, always-on eventual-send support.

### Security Considerations

The replay membrane used by asyncFlow cannot be overridden, so the stopgap `E` option is not available to users of asyncFlow, only to tests that are explicitly using `makeReplayMembraneForTesting`.

### Scaling Considerations

n/a

### Documentation Considerations

n/a

### Testing Considerations

n/a

### Upgrade Considerations

None, provided the version of async flow before this PR has not yet been deployed.

### CODE REVIEW

File: packages/async-flow/src/replay-membrane.js [hunk 1]
```diff
@@ -129,12 +127,30 @@ export const makeReplayMembrane = ({
...
+      // promises are everywhere!
+      // Note: A good place to set a breakpoint, or to uncomment the
+      // `debugger;` line, to work around bundling.
+      // debugger;
+      return watch(h);
+    } else {
+      return h;
+    }
+  };
+
```
Comment (mhofman) [inline packages/async-flow/src/replay-membrane.js]: I don't think we want to re-enable this kludge, do we?

File: packages/async-flow/src/replay-membrane.js [hunk 2]
```diff
@@ -217,193 +233,14 @@ export const makeReplayMembrane = ({
...
-        }
-        default: {
-          // @ts-expect-error TS correctly knows this case would be outside
-          // the type. But that's what we want to check.
-          throw Panic`unexpected outcome kind ${q(outcome.kind)}`;
-        }
+      if (optVerb === undefined) {
+        throw Panic`guest eventual call not yet supported: ${guestTarget}(${b(guestArgs)}) -> ${b(guestReturnedP)}`;
+      } else {
+        throw Panic`guest eventual send not yet supported: ${guestTarget}.${b(optVerb)}(${b(guestArgs)}) -> ${b(guestReturnedP)}`;
```
Comment (mhofman) [inline packages/async-flow/src/replay-membrane.js]: Can't we just replace the `guestHandler` with traps that do this? Possibly configured as an option when preparing the vow tools?

**Review by mhofman — COMMENTED** [review]
It seems that the reverted commit did more than simply enabling E support. I think we'll need to be more surgical. Let's discuss.

File: packages/async-flow/src/replay-membrane.js (hunk 1 above)
Comment (erights) [reply packages/async-flow/src/replay-membrane.js]: We do not. The fact that we are currently running without this kludge means we no longer need it, independent of what our guest-side `E` support is.

File: packages/async-flow/src/replay-membrane.js (hunk 2 above)
Comment (michaelfig) [reply packages/async-flow/src/replay-membrane.js]: I'm a bit confused by what you mean by "replace ... with traps".  Are you talking about Proxy traps?  Or eventual traps?

If eventual traps, that's exactly what the guestHandler already is.

File: packages/async-flow/src/async-flow.js [hunk 3]
```diff
@@ -162,8 +166,9 @@ export const prepareAsyncFlowTools = (outerZone, outerOptions = {}) => {
            * start of a replay.
            *
            * @param {boolean} [eager]
+           * @param {boolean} [eventualSend]
            */
-          restart(eager = startEager) {
+          restart(eager = startEager, eventualSend = enableEventualSend) {
```
Comment (mhofman) [inline packages/async-flow/src/async-flow.js]: I am feeling very unsure about making this a dynamic option.

The problem is that we likely cannot ever support an async flow definition changing its support for eventual send: if the guest code changes to start using eventual send, then most likely it'll fail replaying previous invocations that didn't use it. At least not until we have some sort of support for diverging replay behavior (#10103), and even then I'm suspicious we'd want to encourage such changes.

Interestingly that also means that even when we flip the switch to support eventual send by default, we might want to continue disabling support in existing flow definitions. So far we've always thought we can enable new features that previously were errors, but the fact that async-flows can have both previous incarnation replays to satisfy and new invocations is really scary.

We've talked about having eager wakers that cause the upgrade to fail if they don't replay successfully to mitigate this, but so far have not implemented this.

File: packages/async-flow/test/replay-membrane-eventual.test.js [hunk 4]
```diff
@@ -215,3 +223,22 @@ test.serial('test durable replay-membrane settlement', async t => {
   const zone3 = makeDurableZone(getBaggage(), 'durableRoot');
   return testReplay(t, zone3);
 });
+
+test.serial('test durable disabled eventual send', async t => {
+  annihilate();
+
+  nextLife();
+  const zone1 = makeDurableZone(getBaggage(), 'durableRoot');
+  await testFirstPlay(t, zone1);
```
Comment (mhofman) [inline packages/async-flow/test/replay-membrane-eventual.test.js]: Why are we doing the first play in normal mode?

**Review by mhofman — COMMENTED** [review]
I'm a little confused by the switching behavior of the test. Let's discuss

File: packages/async-flow/test/replay-membrane-eventual.test.js (hunk 4 above)
Comment (michaelfig) [reply packages/async-flow/test/replay-membrane-eventual.test.js]: I was just lazy in updating both async functions.  Both are now tried without and with eventual send support.

File: packages/async-flow/src/async-flow.js (hunk 3 above)
Comment (michaelfig) [reply packages/async-flow/src/async-flow.js]: This option is no longer available except in tests of the replay membrane.

File: packages/async-flow/src/replay-membrane.js [hunk 5]
```diff
@@ -28,12 +29,29 @@ const { fromEntries, defineProperties, assign } = Object;
...
+};
+
+/**
+ * @param {object} arg
+ * @param {LogStore} arg.log
+ * @param {Bijection} arg.bijection
+ * @param {VowTools} arg.vowTools
+ * @param {(vowish: Promise | Vow) => void} arg.watchWake
+ * @param {(problem: Error) => never} arg.panic
+ * @param {boolean} [arg.__eventualSendForTesting] CAVEAT: Only for async-flow tests
```
Comment (mhofman) [inline packages/async-flow/src/replay-membrane.js:46]: I wish there was an easier way to "extend" options while being able to use JSDoc to document the additional option (the only one I'm aware of is multiple `typedef` which is not clean)

**Review by mhofman — APPROVED** [review]
Clean and targeted. Thanks for the updated tests

**Pull request merged**